File: erpnext_dashboard/templates/includes/dashboard/hr/hr.js
/*
	Developer Navdeep
	Email navdeep@blox.ae
*/

frappe.ready(function(){
	erpnext_dashboard.dashboard.HRDashboard = erpnext_dashboard.dashboard.BaseDashboard.extend({
		init: function(args){
			this._super(args);
			$.extend(this, args);
		},
		refresh: function(){
			this._super();
			const filters = this.filters.values;
			this.get_employee_by_departments(filters);
			this.get_absent_and_present(filters);
			this.get_birthday_list(filters);
			this.get_leave_applications(filters);
			this.get_employees_by_year(filters);
			this.get_work_locations(filters);
			this.get_employee_age_group(filters);
			this.get_employee_work_anniversary(filters);
			this.get_new_joinee(filters);
		},
		get_employee_by_departments: function(filters){
			var me = this;
			if(this.employee_by_departments){
				this.employee_by_departments.destroy();
			}
			frappe.call({
				"method": "erpnext_dashboard.www.hr.get_employee_by_departments",
				"args": filters,
				"callback": function(res){
					if(res && res.message){
						var args = {
							"data": res.message,
							"data_field": "total",
							"label_field": "department",
							"parent": $(".emp-by-departments"),
							"chart_type": "doughnut",
							"height": "300"
						};

						this.employee_by_departments = new erpnext_dashboard.dashboard.DoughnutChart(args);
						this.employee_by_departments.render_data();
					}
				}
			})
		},
		get_absent_and_present: function(filters){
			var me = this;
			$(".present-and-absent").empty();
			frappe.call({
				"method": "erpnext_dashboard.www.hr.get_absent_and_present",
				"args": filters,
				"callback": function(res){
					if(res && res.message){
						var $wrapper = $(".present-and-absent");
						for(var i=0;i<res.message.length; i++){
							$(res.message[i]).appendTo($wrapper);
						}
					}
				}
			})
		},
		get_birthday_list: function(filters){
			var me = this;
			$(".birthday-list").empty();
			frappe.call({
				"method": "erpnext_dashboard.www.hr.get_birthday_list",
				"args": filters,
				"callback": function(res){
					if(res && res.message){
						var $wrapper = $(".birthday-list");
						for(var i=0;i<res.message.length; i++){
							$(res.message[i]).appendTo($wrapper);
						}
							
					}
				}
			})	
		},
		get_leave_applications: function(filters){
			var me = this;
			$(".leave-applications").empty();
			frappe.call({
				"method": "erpnext_dashboard.www.hr.get_leave_applications",
				"args": filters,
				"callback": function(res){
					if(res && res.message){
						var $wrapper = $(".leave-applications");
						for(var i=0;i<res.message.length; i++){
							$(res.message[i]).appendTo($wrapper);
						}	
					}
				}
			})
		},
		get_employees_by_year: function(filters){
			if(this.employees_by_year){
				this.employees_by_year.distroy()
			}
			frappe.call({
				"method": "erpnext_dashboard.www.hr.get_employees_by_year",
				"args": filters,
				"callback": function(res){
					if(res && res.message){
						console.log(res.message);
						var args = {
							"data": res.message,
							"data_field": "total",
							"label_field": "year",
							"parent": $(".employees-by-year"),
							"chart_type": "bar",
							"height": "130"
						};
						this.employees_by_years = new erpnext_dashboard.dashboard.BarChart(args);
						this.employees_by_years.render_data();
					}
				}
			});
		},
		get_work_locations: function(filters){
			if(this.work_locations){
				this.work_locations.distroy()
			}
			frappe.call({
				"method": "erpnext_dashboard.www.hr.get_work_locations",
				"args": filters,
				"callback": function(res){
					if(res && res.message){
						console.log(res.message);
						var args = {
							"data": res.message,
							"data_field": "total",
							"label_field": "work_location",
							"parent": $(".work-locations"),
							"chart_type": "pie",
							"height": "130",
						
						};
						this.work_locations = new erpnext_dashboard.dashboard.DoughnutChart(args);
						this.work_locations.render_data();
					}
				}
			});

		},
		get_employee_age_group: function(filters){
			if(this.age_group){
				this.age_group.distroy()
			}
			frappe.call({
				"method": "erpnext_dashboard.www.hr.get_employee_age_group",
				"args": filters,
				"callback": function(res){
					if(res && res.message){
						console.log(res.message);
						var args = {
							"data": res.message,
							"data_field": "total",
							"label_field": "age_group",
							"parent": $(".age-group"),
							"chart_type": "bar",
							"height": "300",
						
						};
						this.age_group = new erpnext_dashboard.dashboard.BarChart(args);
						this.age_group.render_data();
					}
				}
			});
			
		},
		get_employee_work_anniversary: function(filters){
			var me = this;
			$(".employee-work-anniversary").empty();
			frappe.call({
				"method": "erpnext_dashboard.www.hr.get_employee_work_anniversary",
				"args": filters,
				"callback": function(res){
					if(res && res.message){
						var $wrapper = $(".employee-work-anniversary");
						for(var i=0;i<res.message.length; i++){
							$(res.message[i]).appendTo($wrapper);
						}	
					}
				}
			})
		},
		get_new_joinee: function(filters){
			var me = this;
			$(".new-joinee").empty();
			frappe.call({
				"method": "erpnext_dashboard.www.hr.get_new_joinee",
				"args": filters,
				"callback": function(res){
					if(res && res.message){
						var $wrapper = $(".new-joinee");
						for(var i=0;i<res.message.length; i++){
							$(res.message[i]).appendTo($wrapper);
						}	
					}
				}
			})
		},
	});
	var dashboard = new erpnext_dashboard.dashboard.HRDashboard({"filters_fields": erpnext_dashboard.filters});	
	dashboard.init_dashboard();
})

File: erpnext_dashboard/www/hr.py
'''
'''
import frappe
from frappe import _, msgprint, throw
from frappe.utils import nowdate, nowtime, getdate, cint, flt, add_to_date
import json
from erpnext_dashboard.utils import update_random_colors, get_date_diff

no_cache=True

def get_context(context):
	
	filters = json.dumps(get_filters())
	context.update({"filters": filters})
	

def get_filters():
	return[
			{
			"fieldname": "company",
			"fieldtype": "Select",
			"label": _("Company"),
			"class": "company",
			"options": [""]+[val.name for val in frappe.db.get_list("Company")]
		}	
	]

def get_company_condition(company, alias):
	cond = ""
	if company:
		cond += " and {0} = '{1}'".format(alias, company)
	return cond


'''
	GET employee by deparments
'''
@frappe.whitelist()
def get_employee_by_departments(company):
	return update_random_colors(frappe.db.sql("""SELECT e.department, count(e.name) as total, d.color as color 
		from
			`tabEmployee` e inner join `tabDepartment` d 
		on 
			e.department=d.name where e.status='Active' %s 
		group by
			e.department """%(get_company_condition(company, "e.company")), as_dict=True))



'''
	GET list of employees who're absent today
'''
@frappe.whitelist()
def get_absent_and_present(company):
	results = frappe.db.sql("""SELECT e.employee_name as title, image, e.name, a.attendance_date  as sub_title,
			"work_off" as status
		from 
			`tabAttendance` a inner join `tabEmployee` e on a.employee=e.name	
		where 
			a.docstatus=1 and a.attendance_date='%s'  %s 
		order by 
			status limit 4"""%(nowdate(), get_company_condition(company, "a.company")), as_dict=True)

	template = frappe.get_template("templates/includes/dashboard/hr/html/list.html")
	data = []
	for res in results:
		data.append(template.render(res))
	return data


'''
	GET birthday list for all the employee for today's date
'''
@frappe.whitelist()
def get_birthday_list(company):
	results = frappe.db.sql(""" select e.employee_name as title, e.date_of_birth as sub_title, e.image, "event" as status
				from `tabEmployee` e
			
			where 
				e.status='Active' and month(e.date_of_birth)='%s' and day(e.date_of_birth)='%s' 
				%s """%(getdate(nowdate()).month, getdate(nowdate()).day,
					get_company_condition(company, "e.company")), as_dict=True)

	template = frappe.get_template("templates/includes/dashboard/hr/html/list.html")
	data = []
	for res in results:
		data.append(template.render(res))
	return data



'''
	GET last four pending leave applications
'''
@frappe.whitelist()
def get_leave_applications(company):
	
	leaves = frappe.db.sql("""SELECT l.name, l.employee_name, concat_ws(" - ", l.from_date, l.to_date) as date, 
			e.image, l.status, l.leave_type, l.leave_approver, l.description, e.designation,
			total_leave_days
		from 
			`tabLeave Application`  l inner join `tabEmployee` e on l.employee=e.name
		where 
			l.docstatus=0 and l.status not in ('Approved', 'Rejected') and l.from_date >= '%s' %s 
		order by
			l.total_leave_days DESC limit 4""" %(
			nowdate(), get_company_condition(company, "l.company")), as_dict=True)

	data = []
	template  = frappe.get_template("templates/includes/dashboard/hr/html/leave_application.html")
	for leave in leaves:
		data.append(template.render(leave))
	return data


'''
	GET no's of employees joined in past years
'''
@frappe.whitelist()
def get_employees_by_year(company):
	
	return update_random_colors(frappe.db.sql(""" select YEAR(e.date_of_joining) as year, count(e.name) as total 
		from 
			`tabEmployee`  e 
		where 
			status='Active' %s
		group by
			YEAR(e.date_of_joining)
		"""%(get_company_condition(company, "e.company")), as_dict=True))


'''
	GET no's of employees working in different locations
'''
@frappe.whitelist()
def get_work_locations(company):
	return update_random_colors(frappe.db.sql(""" select work_location, count(name) as total from `tabEmployee`
			where 
				status='Active' %s 
			group by
				work_location """%(get_company_condition(company, "company")), as_dict=True))


'''
	GET no's of employees for different age group
'''
@frappe.whitelist()
def get_employee_age_group(company):
	
	results = frappe.db.sql(""" select date_of_joining, date_of_birth from `tabEmployee`
		where
			status='Active' %s """%(get_company_condition(company, "company")), as_dict=True)

	data = [{"15-25":0},{"26-35": 0},{"36-50": 0},{"50+": 0}]

	now = nowdate()
	for res in results:
		diff = abs(get_date_diff(res.date_of_birth, now, "years"))
		if(diff >= 15 and diff <= 25):
			data[0]["15-25"] += 1
		elif(diff >= 26 and diff <= 35):
			data[1]["26-35"] += 1
		elif (diff >= 36 and diff <= 50):
			data[2]["36-50"] += 1
		elif (diff >= 51):
			data[3]["50+"] += 1
	results = []
	for item in data:
		for key, val in item.items():
			results.append({"age_group": key, "total": val})

	return update_random_colors(results)



'''
	Work anniversary for the employees
'''
@frappe.whitelist()
def get_employee_work_anniversary(company):
	
	now = getdate(nowdate())
	results = frappe.db.sql(""" select employee_name as title, designation as sub_title, image from `tabEmployee`
		where
			status = 'Active' and DAY(date_of_joining) = '%s' and 
			MONTH(date_of_joining) ='%s' and DATEDIFF('%s', date_of_joining) >= 365 %s 
		
		"""%(now.day, now.month, now, get_company_condition(company, "company")), as_dict=True)

	template = frappe.get_template("templates/includes/dashboard/hr/html/list.html")
	data = []
	for res in results:
		res.update({"status":""})
		data.append(template.render(res))
	return data


'''
	GET new joiners since last three months to today
'''
@frappe.whitelist()
def get_new_joinee(company):
	
	now = getdate(nowdate())
	after = add_to_date(now, months=-3)
	results = frappe.db.sql(""" select employee_name as title, date_of_joining as sub_title, image from `tabEmployee`
			where
			status = 'Active' and date_of_joining BETWEEN '%s' and '%s' 
			%s """%(after, now, get_company_condition(company, "company")), as_dict=True)
	
	template = frappe.get_template("templates/includes/dashboard/hr/html/list.html")
	data = []
	for res in results:
		res.update({"status":""})
		data.append(template.render(res))
	return data

